### DISCUSSION TITLE
How to implement a Custom CodeActionProvider?

### DISCUSSION THREAD

**Discussion by rhumbertgz** [eclipse-langium/langium#1119] [Q&A] [answered]
Hi, 

Does anyone have implemented a custom CodeActionProvider? 

Is there an example available?

Thanks,
Humberto

**Comment by msujew** [comment, marked as answer]
Hey @rhumbertgz,

we have an example in the [langium grammar language](https://github.com/langium/langium/blob/main/packages/langium/src/grammar/lsp/grammar-code-actions.ts). The idea is just to switch over some attribute of the diagnostic (we use the `code`, but you can use `data` as well) and return a code action where necessary.

**Reply by rhumbertgz** [reply]
Thanks for the pointer @msujew 👍🏼

**Reply by rhumbertgz** [reply]
@msujew I was able to implement my custom CodeActionProvider based on the example that you mentioned, thanks again. But 😀, I noticed that code action icon 💡 is not not displayed.

![Screenshot 2023-07-13 at 09 59 12](https://github.com/langium/langium/assets/10393245/d1b17d51-3ef5-4a84-8230-2ae349da2f2f)

Any suggestions?


You can find the example here:

[Codesanbox](https://codesandbox.io/p/github/rhumbertgz/hello-world/main?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cljzpm8rm000g3t6madexrjpe%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cljzpm8rm000b3t6mbmznt5le%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cljzpm8rm000f3t6mhwsxw2v0%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cljzpm8rm000d3t6m13y4fqe9%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B66.8079280054135%252C33.1920719945865%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cljzpm8rm000b3t6mbmznt5le%2522%253A%257B%2522id%2522%253A%2522cljzpm8rm000b3t6mbmznt5le%2522%252C%2522activeTabId%2522%253A%2522cljzteo8z00a43d62cf4mimul%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Flanguage%252Fhello-world-module.ts%2522%252C%2522id%2522%253A%2522cljzpo6rf019x3t6mmxysghgd%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522id%2522%253A%2522cljzteo8z00a43d62cf4mimul%2522%252C%2522mode%2522%253A%2522temporary%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522cljzpm8rm000d3t6m13y4fqe9%2522%253A%257B%2522id%2522%253A%2522cljzpm8rm000d3t6m13y4fqe9%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522cljzr7uvp018r3t6m7scmzwk3%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%2522%257D%252C%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522serve%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522cljzrtmbq016c3t6mgr6hvuxs%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cljzrtmbq016c3t6mgr6hvuxs%2522%257D%252C%2522cljzpm8rm000f3t6mhwsxw2v0%2522%253A%257B%2522id%2522%253A%2522cljzpm8rm000f3t6mhwsxw2v0%2522%252C%2522activeTabId%2522%253A%2522cljzrnl7r00kn3t6m70b4aths%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cljzpm8rm000e3t6mtftxitnj%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cljzr70qw0028fmil8q980e5a%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522serve%2522%252C%2522id%2522%253A%2522cljzrnl7r00kn3t6m70b4aths%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A19.153196839080465%257D)

[Github repository](https://github.com/rhumbertgz/hello-world)

**Reply by rhumbertgz** [reply]
Using the react component the quick action is displayed as aspected. It seems I missing some setting here.

![image](https://github.com/langium/langium/assets/10393245/3dedba13-9597-41cc-be90-408840e28e90)

**Reply by spoenemann** [reply]
Does it make a difference if you set `isPreferred` to `true`?

**Reply by rhumbertgz** [reply]
@spoenemann I had already the property `isPreferred` to `true`. I basically copy-paste my action provider. I tried to disable some properties of the `UserConfig` looking for the one activating the icons, but nothing worked. At least, it works perfect on my real project :). Thanks anyways!
